### PR TITLE
Extend `responseDataPath` to support extracting objects inside arrays.

### PR DESCRIPTION
**Pull request by na9da**
### What this PR does

Required to fix https://github.com/TerriaJS/vic-digital-twin/issues/457

Currently the `responseDataPath` trait of `ApiTableCatalogItem` (implemented using [lodash.get](https://lodash.com/docs/4.17.15#get) function) supports the following syntaxes:
- `a.b.c` for extracting object under `a > b > c`
- `a.b[1].c` for extracting a **single** `c` object under `a > b[1]`

This PR adds support for the following path:
- `a.b[].c` for extracting all `c`s under `a > b[]` array.

The syntax is similar to the `jq` [syntax for traversing arrays](https://stedolan.github.io/jq/manual/#Array/ObjectValueIterator:.[]).

### Test me

- Test this [CI link](http://ci.terria.io/response-data-path-array/#share=s-vcHZ8Y1x6JGGz9MnHOnbes4NnlW&clean&https://gist.githubusercontent.com/na9da/8bd6fcdd5b1894227a42028dd7adba03/raw/0384ba0b019fafb1891e99a99a017f1ce0d1e793/melbourne-pedestrian-traffic.json) which is using this [catalog definition](https://gist.github.com/na9da/8bd6fcdd5b1894227a42028dd7adba03#file-melbourne-pedestrian-traffic-json-L7-L12) with the new `responseDataPath` syntax.

### Checklist

- [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
- [ ] I've updated relevant documentation in `doc/`.
- [ ] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts [hunk 1]
```diff
@@ -253,4 +253,28 @@ export class ApiTableCatalogItem extends AutoRefreshingMixin(
...
 
+/**
+ * Return the value at json path of the data object.
+ *
+ * This works exactly like the lodash.get() function but adds support for
+ * traversing array objects.  For eg, the lodash.get() does not support a path
+ * like: `a.users[].name`, but this function will correctly return a `{name}[]`
+ * array if they exist. The particular syntax for array traversal
+ * is borrowed from `jq` CLI tool.
+ */
```
Comment (staffordsmith83) [inline lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts:265]: Can we define the type of `data` as an object where we dont know the properties maybe?

File: lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts [hunk 2]
```diff
@@ -253,4 +253,28 @@ export class ApiTableCatalogItem extends AutoRefreshingMixin(
...
+function getResponseDataPath(data: any, jsonPath: string) {
+  // Split the path at `[].` or `[]`
+  const pathSegments = jsonPath.split(/\[\]\.?/);
+  const getPath = (data: any, path: string) =>
```
Comment (staffordsmith83) [inline lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts]: Same here, define type of data at all, even loosely?

File: lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts [hunk 3]
```diff
@@ -253,4 +253,28 @@ export class ApiTableCatalogItem extends AutoRefreshingMixin(
...
+    path === ""
+      ? data
+      : Array.isArray(data)
+      ? _map(data, path)
+      : _get(data, path);
+  return pathSegments.reduce(
```
Comment (staffordsmith83) [inline lib/Models/Catalog/CatalogItems/ApiTableCatalogItem.ts:275]: This is a really neat algorithm nice one

File: test/Models/Catalog/CatalogItems/ApiTableCatalogItemSpec.ts [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            {
+              title: "longitude",
+              name: "longitude"
+            }
+          ]
+        });
+      });
+    });
+
+    it("can extract value from a nested object", async function () {
```
Comment (staffordsmith83) [inline test/Models/Catalog/CatalogItems/ApiTableCatalogItemSpec.ts:132]: Great tests

**Review by staffordsmith83 — COMMENTED** [review]
Looks great just going to test with other ApiTable CatalogItems - One change requested: can we do anything about the `any` type for data.

**Review by staffordsmith83 — CHANGES_REQUESTED** [review]
Just look at `any`

**Comment by na9da** [comment]
Thanks for the review @staffordsmith83. As discussed offline I have replaced `any` with `JsonValue` which is a better fit in this case.

**Review by staffordsmith83 — APPROVED** [review]
Looks good!

**Pull request merged**